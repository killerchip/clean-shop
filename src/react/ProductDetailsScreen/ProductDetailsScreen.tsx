import { observer } from "mobx-react-lite";
import { Stack, useLocalSearchParams } from "expo-router";
import { useCreateProductDetailsScreenPresenter } from "../../presentation/ProductDetailsScreen.presenter";
import { ScrollView, StyleSheet, View } from "react-native";
import { Image } from "expo-image";
import { Text, Button, useTheme } from "react-native-paper";
import { CartIcon } from "../components/CartIcon";

const blurhash =
  "|rF?hV%2WCj[ayj[a|j[az_NaeWBj@ayfRayfQfQM{M|azj[azf6fQfQfQIpWXofj[ayj[j[fQayWCoeoeaya}j[ayfQa{oLj?j[WVj[ayayj[fQoff7azayj[ayj[j[ayofayayayj[fQj[ayayj[ayfjj[j[ayjuayj[";

export const ProductDetailsScreen = observer(function ProductsDetailsScreen() {
  const { id } = useLocalSearchParams<{ id: string }>();
  const presenter = useCreateProductDetailsScreenPresenter(Number(id));

  const onCartPress = () => presenter.addProductToCart();
  const theme = useTheme();
  const { itemsInCart, product } = presenter;

  return (
    <>
      <Stack.Screen
        options={{
          title: product?.title,
          headerRight: () => <CartIcon itemsNumber={itemsInCart} />,
        }}
      />
      <ScrollView style={styles.container}>
        <Image
          source={product?.image}
          style={styles.image}
          transition={100}
          placeholder={{ blurhash }}
        />
        <View style={styles.textContainer}>
          <Text style={styles.title} variant="headlineMedium">
            {product?.title}
          </Text>
          <View style={styles.actionLine}>
            <Text style={[styles.price, { color: theme.colors.primary }]}>
              ${product?.price}
            </Text>
            <Button mode="contained" onPress={onCartPress} icon="cart">
              Add To Cart
            </Button>
          </View>
          <Text style={styles.description}>{product?.description}</Text>
        </View>
      </ScrollView>
    </>
  );
});

// TODO: This should be a wrapper component
const styles = StyleSheet.create({
  container: {
    flex: 1,
  },
  image: { width: "100%", height: 250 },
  textContainer: {
    paddingHorizontal: 20,
  },
  title: {
    textAlign: "center",
    marginTop: 6,
    fontWeight: "bold",
    paddingVertical: 10,
  },
  description: {
    fontSize: 16,
    fontStyle: "italic",
    paddingTop: 10,
  },
  price: {
    fontSize: 25,
    fontWeight: "bold",
  },
  actionLine: {
    flexDirection: "row",
    justifyContent: "space-between",
    paddingVertical: 10,
    alignItems: "center",
  },
});
