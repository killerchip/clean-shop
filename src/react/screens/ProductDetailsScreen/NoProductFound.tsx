import styled from "styled-components/native";
import { View } from "react-native";
import { Text } from "react-native-paper";

export function NoProductFound() {
  return (
    <OopsContainer>
      <Text>Oops! No product found</Text>
    </OopsContainer>
  );
}

const OopsContainer = styled(View)`
  flex: 1;
  justify-content: center;
  align-items: baseline;
`;
