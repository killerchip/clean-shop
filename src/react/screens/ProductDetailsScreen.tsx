import { observer } from "mobx-react-lite";
import { Stack, useLocalSearchParams } from "expo-router";
import { useCreateProductDetailsScreenPresenter } from "../../presentation/ProductDetailsScreen.presenter";
import { Button, StyleSheet, Text, View } from "react-native";

export const ProductDetailsScreen = observer(function ProductsDetailsScreen() {
  const { id } = useLocalSearchParams<{ id: string }>();
  const presenter = useCreateProductDetailsScreenPresenter(Number(id));

  const onCartPress = () => presenter.addProductToCart();

  return (
    <>
      <Stack.Screen options={{ title: presenter.product?.title }} />
      {presenter.displayAddToCartButton && (
        <Button
          title={`Add To Cart: ${presenter.itemsInCart}`}
          onPress={onCartPress}
        />
      )}
      <View style={styles.container}>
        <Text>{JSON.stringify(presenter.product, null, 2)}</Text>
      </View>
    </>
  );
});

// TODO: This should be a wrapper component
const styles = StyleSheet.create({
  container: {
    flex: 1,
  },
});
