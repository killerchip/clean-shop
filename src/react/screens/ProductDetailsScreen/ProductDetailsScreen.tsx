import { observer } from "mobx-react-lite";
import { Stack, useLocalSearchParams } from "expo-router";
import { useCreateProductDetailsScreenPresenter } from "../../../presentation/ProductDetailsScreen.presenter";
import { View } from "react-native";
import { Image } from "expo-image";
import { Text, Button } from "react-native-paper";
import { CartIcon } from "../../components/CartIcon";
import { PageScrollView } from "../../components/atoms/PageScrollView";
import { PriceTagMedium } from "../../components/atoms/PriceTags";
import styled from "styled-components/native";

const blurhash =
  "|rF?hV%2WCj[ayj[a|j[az_NaeWBj@ayfRayfQfQM{M|azj[azf6fQfQfQIpWXofj[ayj[j[fQayWCoeoeaya}j[ayfQa{oLj?j[WVj[ayayj[fQoff7azayj[ayj[j[ayofayayayj[fQj[ayayj[ayfjj[j[ayjuayj[";

export const ProductDetailsScreen = observer(function ProductsDetailsScreen() {
  const { id } = useLocalSearchParams<{ id: string }>();
  const presenter = useCreateProductDetailsScreenPresenter(Number(id));

  const onCartPress = () => presenter.addProductToCart();
  const { itemsInCart, product } = presenter;

  return (
    <>
      <Stack.Screen
        options={{
          title: product?.title,
          headerBackTitleVisible: false,
          headerRight: () => <CartIcon itemsNumber={itemsInCart} />,
        }}
      />
      <PageScrollView>
        <ProductDetailsImage
          source={product?.image}
          transition={100}
          placeholder={{ blurhash }}
        />
        <TextContainer>
          <Title variant="headlineMedium">{product?.title}</Title>
          <ActionLine>
            <PriceTagMedium>${product?.price}</PriceTagMedium>
            <Button mode="contained" onPress={onCartPress} icon="cart">
              Add To Cart
            </Button>
          </ActionLine>
          <Description>{product?.description}</Description>
        </TextContainer>
      </PageScrollView>
    </>
  );
});

const ProductDetailsImage = styled(Image)`
  width: 100%;
  height: 250px;
`;

const Title = styled(Text)`
  text-align: center;
  margin-top: 6px;
  font-weight: bold;
  padding-top: 10px;
  padding-bottom: 10px;
`;

const ActionLine = styled(View)`
  flex-direction: row;
  justify-content: space-between;
  padding-top: 10px;
  padding-bottom: 10px;
  align-items: center;
`;

const Description = styled(Text)`
  font-size: 16px;
  font-style: italic;
  padding-top: 10px;
`;

const TextContainer = styled(View)`
  padding-left: 20px;
  padding-right: 20px;
`;
