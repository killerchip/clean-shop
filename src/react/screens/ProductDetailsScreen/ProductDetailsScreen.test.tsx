import "reflect-metadata";
import { fireEvent, screen } from "@testing-library/react-native";
import { ProductDetailsScreen } from "./ProductDetailsScreen";
import { useCreateProductDetailsScreenPresenter } from "@/presentation/ProductDetailsScreen.presenter";
import { renderRouter } from "expo-router/testing-library";
import { useLocalSearchParams } from "expo-router";

// TODO: fix the warning error on console. https://github.com/expo/expo/issues/28831

jest.mock("../../../presentation/ProductDetailsScreen.presenter");
// mock non existing modules
jest.mock("react-native-reanimated", () => null, {
  virtual: true,
});

jest.mock("expo-router", () => {
  const original = jest.requireActual("expo-router"); // Step 2.
  return {
    ...original,
    useLocalSearchParams: jest.fn(),
  };
});

describe("ProductDetailsScreen", () => {
  const MockComponent = jest.fn(() => <ProductDetailsScreen />);
  const mockAddProductToCart = jest.fn();
  const mockProduct = {
    id: 1,
    title: "Test Product",
    description: "Test Description",
    price: "100",
    image: "test-image-url",
  };

  beforeEach(() => {
    (useCreateProductDetailsScreenPresenter as jest.Mock).mockReturnValue({
      addProductToCart: mockAddProductToCart,
      itemsInCart: 1,
      product: mockProduct,
    });
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it("instantiates the presenter with the correct id", () => {
    (useLocalSearchParams as jest.Mock).mockImplementation(() => ({ id: "1" }));
    renderRouter(
      {
        "/product/1": MockComponent,
      },
      { initialUrl: "/product/1" },
    );

    expect(useCreateProductDetailsScreenPresenter).toHaveBeenCalledWith(1);
  });

  it("displays product details when product is available", () => {
    renderRouter(
      {
        "/product/1": MockComponent,
      },
      { initialUrl: "/product/1" },
    );

    expect(screen.getByText(mockProduct.title)).toBeTruthy();
    expect(screen.getByText(mockProduct.description)).toBeTruthy();
    expect(screen.getByText(`$${mockProduct.price}`)).toBeTruthy();
  });

  it("displays passes the snapshot test", () => {
    renderRouter(
      {
        index: MockComponent,
        "/product/1": MockComponent,
      },
      { initialUrl: "/product/1" },
    );
    expect(screen.toJSON()).toMatchSnapshot();
  });

  it("calls addProductToCart when add to cart button is pressed", () => {
    renderRouter(
      {
        index: MockComponent,
        "/product/1": MockComponent,
      },
      { initialUrl: "/product/1" },
    );

    fireEvent.press(screen.getByText("Add To Cart"));

    expect(mockAddProductToCart).toHaveBeenCalled();
  });

  it.skip("updates cart icon with the correct number of items in cart", () => {
    // This one does not work as the renderRouter does not include a header
    // The correctness of the badge is tested in the CartIcon component
    // and in the integration test of the presenter
    // To test the whole screen, probably need to use E2E testing
  });

  it("displays an error message when product is not available", () => {
    (useCreateProductDetailsScreenPresenter as jest.Mock).mockReturnValue({
      addProductToCart: mockAddProductToCart,
      itemsInCart: 1,
      product: undefined,
    });

    renderRouter(
      {
        index: MockComponent,
        "/product/1": MockComponent,
      },
      { initialUrl: "/product/1" },
    );

    expect(screen.getByText("Oops! No product found")).toBeTruthy();
  });
});
