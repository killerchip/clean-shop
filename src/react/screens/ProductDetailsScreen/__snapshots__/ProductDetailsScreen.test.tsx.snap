// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductDetailsScreen displays passes the snapshot test 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    [
      {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <RNCSafeAreaView
    edges={
      {
        "bottom": "additive",
        "left": "additive",
        "right": "additive",
        "top": "additive",
      }
    }
    style={
      {
        "flex": 1,
      }
    }
  >
    <RCTScrollView
      style={
        {
          "flexBasis": 0,
          "flexGrow": 1,
          "flexShrink": 1,
        }
      }
    >
      <View>
        <ViewManagerAdapter_ExpoImage
          contentFit="cover"
          contentPosition={
            {
              "left": "50%",
              "top": "50%",
            }
          }
          height={250}
          nativeViewRef={"[React.ref]"}
          onError={[Function]}
          onLoad={[Function]}
          onLoadStart={[Function]}
          onProgress={[Function]}
          placeholder={
            [
              {
                "height": 16,
                "uri": "blurhash:/%7CrF%3FhV%252WCj%5Bayj%5Ba%7Cj%5Baz_NaeWBj@ayfRayfQfQM%7BM%7Cazj%5Bazf6fQfQfQIpWXofj%5Bayj%5Bj%5BfQayWCoeoeaya%7Dj%5BayfQa%7BoLj%3Fj%5BWVj%5Bayayj%5BfQoff7azayj%5Bayj%5Bj%5Bayofayayayj%5BfQj%5Bayayj%5Bayfjj%5Bj%5Bayjuayj%5B",
                "width": 16,
              },
            ]
          }
          source={
            [
              {
                "uri": "test-image-url",
              },
            ]
          }
          style={
            {
              "height": 250,
              "width": "100%",
            }
          }
          transition={
            {
              "duration": 100,
            }
          }
          width="100%"
        />
        <View
          style={
            {
              "paddingLeft": 20,
              "paddingRight": 20,
            }
          }
        >
          <Text
            style={
              [
                {
                  "textAlign": "left",
                },
                {
                  "color": "rgba(28, 27, 31, 1)",
                  "writingDirection": "ltr",
                },
                [
                  {
                    "fontFamily": "System",
                    "fontSize": 28,
                    "fontWeight": "400",
                    "letterSpacing": 0,
                    "lineHeight": 36,
                  },
                  {
                    "fontWeight": "bold",
                    "marginTop": 6,
                    "paddingBottom": 10,
                    "paddingTop": 10,
                    "textAlign": "center",
                  },
                ],
              ]
            }
          >
            Test Product
          </Text>
          <View
            style={
              {
                "alignItems": "center",
                "flexDirection": "row",
                "justifyContent": "space-between",
                "paddingBottom": 10,
                "paddingTop": 10,
              }
            }
          >
            <Text
              style={
                [
                  {
                    "textAlign": "left",
                  },
                  {
                    "color": "rgba(28, 27, 31, 1)",
                    "fontFamily": "System",
                    "fontWeight": "400",
                    "letterSpacing": 0,
                  },
                  {
                    "writingDirection": "ltr",
                  },
                  {
                    "color": "rgba(103, 80, 164, 1)",
                    "fontSize": 25,
                  },
                ]
              }
            >
              $
              100
            </Text>
            <View
              collapsable={false}
              style={
                {
                  "backgroundColor": "rgba(103, 80, 164, 1)",
                  "borderRadius": 20,
                  "shadowColor": "#000",
                  "shadowOffset": {
                    "height": 0,
                    "width": 0,
                  },
                  "shadowOpacity": 0,
                  "shadowRadius": 0,
                }
              }
              testID="button-container-outer-layer"
            >
              <View
                collapsable={false}
                style={
                  {
                    "backgroundColor": "rgba(103, 80, 164, 1)",
                    "borderColor": "transparent",
                    "borderRadius": 20,
                    "borderStyle": "solid",
                    "borderWidth": 0,
                    "flex": undefined,
                    "minWidth": 64,
                    "shadowColor": "#000",
                    "shadowOffset": {
                      "height": 0,
                      "width": 0,
                    },
                    "shadowOpacity": 0,
                    "shadowRadius": 0,
                  }
                }
                testID="button-container"
              >
                <View
                  accessibilityRole="button"
                  accessibilityState={
                    {
                      "busy": undefined,
                      "checked": undefined,
                      "disabled": false,
                      "expanded": undefined,
                      "selected": undefined,
                    }
                  }
                  accessibilityValue={
                    {
                      "max": undefined,
                      "min": undefined,
                      "now": undefined,
                      "text": undefined,
                    }
                  }
                  accessible={true}
                  collapsable={false}
                  focusable={true}
                  onBlur={[Function]}
                  onClick={[Function]}
                  onFocus={[Function]}
                  onResponderGrant={[Function]}
                  onResponderMove={[Function]}
                  onResponderRelease={[Function]}
                  onResponderTerminate={[Function]}
                  onResponderTerminationRequest={[Function]}
                  onStartShouldSetResponder={[Function]}
                  style={
                    [
                      {
                        "overflow": "hidden",
                      },
                      {
                        "borderRadius": 20,
                      },
                    ]
                  }
                  testID="button"
                >
                  <View
                    style={
                      [
                        {
                          "alignItems": "center",
                          "flexDirection": "row",
                          "justifyContent": "center",
                        },
                        undefined,
                      ]
                    }
                  >
                    <View
                      style={
                        [
                          {
                            "marginLeft": 12,
                            "marginRight": -4,
                          },
                          {
                            "marginLeft": 16,
                            "marginRight": -16,
                          },
                          false,
                        ]
                      }
                      testID="button-icon-container"
                    >
                      <Text
                        accessibilityElementsHidden={true}
                        allowFontScaling={false}
                        importantForAccessibility="no-hide-descendants"
                        pointerEvents="none"
                        selectable={false}
                        style={
                          [
                            {
                              "color": "rgba(255, 255, 255, 1)",
                              "fontSize": 18,
                            },
                            [
                              {
                                "lineHeight": 18,
                                "transform": [
                                  {
                                    "scaleX": 1,
                                  },
                                ],
                              },
                              {
                                "backgroundColor": "transparent",
                              },
                            ],
                            {
                              "fontFamily": "material-community",
                              "fontStyle": "normal",
                              "fontWeight": "normal",
                            },
                            {},
                          ]
                        }
                      >
                        Û∞Ñê
                      </Text>
                    </View>
                    <Text
                      numberOfLines={1}
                      selectable={false}
                      style={
                        [
                          {
                            "textAlign": "left",
                          },
                          {
                            "color": "rgba(28, 27, 31, 1)",
                            "writingDirection": "ltr",
                          },
                          [
                            {
                              "fontFamily": "System",
                              "fontSize": 14,
                              "fontWeight": "500",
                              "letterSpacing": 0.1,
                              "lineHeight": 20,
                            },
                            [
                              {
                                "marginHorizontal": 16,
                                "marginVertical": 9,
                                "textAlign": "center",
                              },
                              false,
                              {
                                "marginHorizontal": 24,
                                "marginVertical": 10,
                              },
                              undefined,
                              false,
                              {
                                "color": "rgba(255, 255, 255, 1)",
                                "fontFamily": "System",
                                "fontSize": 14,
                                "fontWeight": "500",
                                "letterSpacing": 0.1,
                                "lineHeight": 20,
                              },
                              undefined,
                            ],
                          ],
                        ]
                      }
                      testID="button-text"
                    >
                      Add To Cart
                    </Text>
                  </View>
                </View>
              </View>
            </View>
          </View>
          <Text
            style={
              [
                {
                  "textAlign": "left",
                },
                {
                  "color": "rgba(28, 27, 31, 1)",
                  "fontFamily": "System",
                  "fontWeight": "400",
                  "letterSpacing": 0,
                },
                {
                  "writingDirection": "ltr",
                },
                {
                  "fontSize": 16,
                  "fontStyle": "italic",
                  "paddingTop": 10,
                },
              ]
            }
          >
            Test Description
          </Text>
        </View>
      </View>
    </RCTScrollView>
  </RNCSafeAreaView>
</RNCSafeAreaProvider>
`;
